wikimedia/wikipedia-ios
# Variant-aware article summary keys

This is work towards the Chinese variants feature https://phabricator.wikimedia.org/T195265
It is not the entire feature/ticket.

Each commit is self-contained it is probably easiest to review them individually.

When processing fetched articles and article summaries, dictionaries are used to check for updates, match summaries to articles, etc. They keys currently used are wmf_databaseKey values, which do not take variants into account.

This PR updates all of those call stacks to use a compound key value. 

### Notes
**Commit 1**
Rename the existing WMFArticleTemporaryCacheKey to WMFInMemoryURLKey and move definition to NSURL category to reflect its more general usage. Update call sites.

**Commit 2**
Rename variant property to languageVariantCode to match the URL property naming. Mark the languageVariantCode nullable to correctly indicate it can be nil. Update call sites.

**Commit 3**
Add convenience methods and properties for WMFInMemoryURLKey to round trip to URL and for URL to return an in memory key. Will be used in upcoming commits.

**Commit 4**
Add convenience methods and properties in WMFArticle+Extensions which will make the intent of call sites when used more clear.

**Commit 5**
The main event! Update ArticleFetcher methods to use in-memory key instead of database key. Change all call stacks to use the new type. Move off of 'fetchWtihKey:' article fetching methods which are not variant-aware.

**Commit 6**
Update predicate in SummaryExtensions to take key and variant into account. Since this is a more complex predicate, add it as a method in WMFArticle+Extensions. It will be used more as more subsystems become variant-aware.

**Commit 7**
Remove shim 'fetchWtihKey:' article fetching methods that are no longer used after these changes. 

**Commit 8**
Add a partial fix for propagating language variant to ArticleSummary instances. This addresses the issue where navigating from the explore feed to an article using a variant and back to the explore feed would cause the incorrect variant to be displayed. It also addresses an issue where the other changes in this PR would cause a blank feed card after the same navigation pattern.

A more generalized approach for propagating language variants to Codable types will be in a future PR.

### Test Steps
1. With language variant feature turned off, app behavior should not change.
2. With language variants turned on, you can now navigate from the explore feed to an article and back and the correct variant information displays in the explore feed.
